input_rc: take command from the last input_rc received

NicolasM0 opened the pull request:
Before this commit, if there are multiple instances of input_rc topic, only the first one is used.
We have 2 companions for redundancy, if the first one fail the second one is not able to take over control because commands are published in another input_rc instance.

This commit allow to use the instance with the last rc command received (with a maximum of thee instances).

dagar commented:
I'm wondering if this needs some level of configuration to prevent potential accidents or confusion.

In particular switching between RC input and manual control via mavlink is a recurring request, but so far I've pushed back until we have a safe explicit mechanism for changing the primary input.

At at a minimum for this PR how about a new parameter that controls how multiple are handled? The default could be to latch to the first available (basically the current behaviour), then you could an option to enable what you want here (use most recent available).

NicolasM0 commented:
@dagar Thank you for the review.
To give you some background, our architecture constrains me to use manual input from two companions. It worked on 1.6.5 and we try to update to a newer version.
Adding a new parameter is ok for me. I will do it in this PR

dagar commented:
> @dagar Thank you for the review.
> To give you some background, our architecture constrains me to use manual input from two companions. It worked on 1.6.5 and we try to update to a newer version.
> Adding a new parameter is ok for me. I will do it in this PR

Thanks, knowing the full use case is helpful. The behaviour was changed because of possible confusion when receiving both RC input and Mavlink MANUAL_CONTROL. In that case it was possible to lose RC, not trigger the failsafe, and not be ready to takeover manually.

NicolasM0 commented on src/modules/rc_update/params.c:
```diff
@@ -2251,3 +2251,19 @@ PARAM_DEFINE_FLOAT(RC_FLT_SMP_RATE, 50.0f);
+ * Warning, in this case, if radio control is lost and RC commands are recived via mavlink (see COM_RC_IN_MODE=2),
+ * RC lost fail safe will not be triggered.
+ *
+ * @group Commander
```
I added this parameter to the commander group even if it is used in rc_update because I think it makes more sens. Is it ok or should I move it to another group (for example Radio Calibration)

NicolasM0 commented on src/modules/rc_update/rc_update.cpp:
```diff
@@ -340,8 +345,23 @@ RCUpdate::Run()
+	 * find the one that triggered the callback */
+	if (!input_rc_copy_ok && _param_com_rc_multiple_inputs.get() > 0) {
+		for (int i = 1; i < MAX_INPUT_RC; i++) {
+			if (_input_rc_subs[i].updated()) {
```
Is it ok to check if it is updated or should I check which rc_input.timestamp is the last ?

NicolasM0 commented:
If COM_RC_IN_MODE is set to "Virtual RC by Joystick" and a rc and virtual joysticks are used, they will share the same channel mapping whereas they should be different.

Is it a blocking issue ?

Another issues is that if two inputs are used at the same time, drone will switch from one command to another resulting in an controlled behavior. To fix that, I can replace the 'updated()" check by checking a if no command have been received since COM_RC_LOSS_T before checking following instances. This will a priority to the first instance.

NicolasM0 commented:
@MaEtUgR following our discussion in slack, I added a comment here. what do you think ?

NicolasM0 commented:
I fixed the issue of conflicts when multiple inputs are updated at the same time so I rebased my commit and force pushed it.
Is it possible to reopen the PR and have a feedback ?

I also added a warning for the common RC mapping issue.

Thanks

TSC21 commented:
@NicolasM0 format is failing. Please run `make format`.

NicolasM0 commented:
> @NicolasM0 format check is failing. Please run `make format`.

Does not seems to be linked to my changes
Fixed with rebase

NicolasM0 commented:
@dagar changed as discussed on slack.

dagar commented:
@NicolasM0 sorry, this one slipped through. If you're still interested in this can you rebase and we'll get it in?

NicolasM0 commented:
> @NicolasM0 sorry, this one slipped through. If you're still interested in this can you rebase and we'll get it in?

Yes, I still used it on our local branch so if it could be merge, it would be great.
